### DISCUSSION TITLE
How to use storage_secret when calling the @page function from another module?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1289] [Q&A] [answered]
### Question

Hi All,

I am trying to create a login system, with the code split up into multiple code files, like let's say I have the following:

1. app.py -- this is the main file where I am initiating the app and I have the ```ui.run(storage_secret = '1234')```
2. sign_in.py -- this is the login script, where I check for username and password and if all okay, I redirect them to homepage
3. sign_up.py -- for new users to sign up
4. home.py -- this is the page they will see once they login successfully, I need to pass the ```app.storage.user``` dict to this page 

So in my sign_in.py, in the ```submit``` button, I am using this function below to check for correct login
```python
    def login():
        # Check for username, password errors
        # skipping this code here 

        # Login
        ui.open(homepage)
```

and then the homepage function is like this below:

```python
    @ui.page('/home')
    def homepage():

        # Initiate session state
        app.storage.user['username'] = username.value     # this is coming from the user input

        # Redirect to homepage
        home.home(app.storage.user['username'])
```

So when I run this code from the main ```app.py``` , I get an error saying 
```
RuntimeError: app.storage.user needs a storage_secret passed in ui.run()
```

So I guess if I pass the storage_secret from the app.py, it is not working, is it? Does it need to be on the sign_in.py where I am running the ```homepage()``` function?

Kindly advise me on how to resolve this please. I am new to the session state concepts, and basically trying to understand how to create and pass these dict values across my scripts from one file to another. 

Regards,
Anindya

**Comment by rodja** [comment]
The storage secret is set globally and is therefore available in all files. There must be something else causing the error. Can you try to debug it by reducing your code to the bare minimum which reproduces the error? Like running only a line like 

```py
app.storage.user['username'] = 'test'
```

in your login function, or as replacement of the submit button? Does it work if the test code is run in the `app.py` file? Also try renaming `app.py` to `main.py`.

**Reply by Anindya088** [reply]
Hi @rodja , thanks for the explanation. I did some further tests and everything is matching with what you are saying.

For ex: I minimized everything to 2 files, because essentially I just wanted to check if the page function can be called from another code file:
So I made main.py and homepage.py

main.py
```python
from nicegui import ui, app
import homepage

ui.link('homepage', homepage.homepage)

ui.run(storage_secret='1234')
```

homepage.py
```python
from nicegui import ui, app

@ui.page('/home')
def homepage():
    app.storage.user['username'] = 'test'
    ui.label(app.storage.user['username'])
```

When I simply run this, this works perfectly well, and I see the word 'test' in my browser at http://localhost:8080/home
So it works just fine.

Now I need to find out what is wrong with my code, surely it's not the storage_secret. 
Let me debug more and try to find out the source error.

Thanks for the help on this.
Anindya

**Reply by Anindya088** [reply]
Hi @rodja , I managed to debug my previous code and make it work, so thanks for the suggestions before. 
Now I am having another challenge, which is that... how to redirect users back to login page if they are trying to access some other pages.

So taking the same minimal example from my previous reply, what I am trying to do is this:

main.py -- the main file I am calling
```python
from nicegui import ui, app
import homepage

ui.link('homepage', homepage.homepage)

ui.run(storage_secret='1234') 
```

homepage.py -- this is where the user will be routed if he is logged in, otherwise back to login page
```python
from nicegui import ui, app

@ui.page('/home')
def homepage():

    # Simulate login status
    app.storage.user['username'] = 'test1'
    app.storage.user['logged_on'] = False

    # If login then allow them to see page
    if app.storage.user['logged_on']:
        ui.label('You have logged in')
    # Otherwise route them back to login page
    else:
        ui.open(login)



@ui.page('/login')
def login():
    ui.label('This is the login page')
```

So, when I make the ```logged_on = True```, then it's fine, the home page loads with the label correctly.
But when I make the ```logged_on = False```, then also the home page loads, just that it is blank now.

So that means, the storage variable is working correctly, and the label is not displayed. But then I thought  ```ui.open(login)``` would redirect the user to the ```login``` page.  But that is not happening.

Can you please advise what should be the correct approach here? Basically the page function ```homepage``` I want to make such that if the storage variable is false then instead of displaying that page it goes back to ```login``` page.

Please let me know what is the right way to do this. Thanks a lot for the help.

Regards,
Anindya

**Comment by rodja** [comment, marked as answer]
Have you seen our [auth example](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py)? We use a `RedirectResponse` instead of `ui.open` which only works after the websocket connection has been established:

```py
from nicegui import ui, Client

@ui.page('/')
async def index(client: Client):
    await client.connected()
    ui.open('/test')

@ui.page('/test')
def index():
    ui.label('test')

ui.run()
```

**Reply by Anindya088** [reply]
Hi @rodja , 

Thanks a lot for the explanation, so now I understand that ```ui.open``` cannot be used in this case, did not know that.

Actually I have gone through the auth example before and took quite a few ideas from there. But because I am not at all familiar with FastAPI and routing related things at all, so I did not dare to try it :) I thought I will just stick to the NiceGUI components, haha 😄 

But let me go back and read through it again, I will try to implement the  ```RedirectResponse``` thing that you advised.

Thanks a lot for the help.
Anindya
